flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java

```diff
@@ -54,39 +62,157 @@
+				LOG.warn("not processing any records in IDLE state");
+				return false;
+			}
+		},
+		/**
+		 * A message is enqueued to process split, but no split is opened.
+		 */
+		OPENING { // the split was added and message to itself was enqueued to process it
+			@Override
+			public boolean prepareToProcessRecord(ContinuousFileReaderOperator<?> op) throws IOException {
+				if (op.splits.isEmpty()) {
+					op.switchState(ReaderState.IDLE);
+					return false;
+				} else {
+					op.loadSplit(op.splits.poll());
+					op.switchState(ReaderState.READING);
+					return true;
+				}
+			}
+		},
+		/**
+		 * A message is enqueued to process split and its processing was started.
+		 */
+		READING {
+			@Override
+			public boolean prepareToProcessRecord(ContinuousFileReaderOperator<?> op) {
+				return true;
+			}
 
-	private transient Object checkpointLock;
+			@Override
+			public void onNoMoreData(ContinuousFileReaderOperator<?> op) {
+				op.switchState(ReaderState.IDLE);
+			}
+		},
+		/**
+		 * {@link #close()} was called but unprocessed data (records and splits) remains and needs to be processed.
+		 * {@link #close()} caller is blocked.
+		 */
+		CLOSING {
+			@Override
+			public boolean prepareToProcessRecord(ContinuousFileReaderOperator<?> op) throws IOException {
+				if (op.currentSplit == null && !op.splits.isEmpty()) {
+					op.loadSplit(op.splits.poll());
+				}
+				return true;
+			}
 
-	private transient SplitReader<OUT> reader;
-	private transient SourceFunction.SourceContext<OUT> readerContext;
+			@Override
+			public void onNoMoreData(ContinuousFileReaderOperator<?> op) {
+				// need one more mail to unblock possible yield() in close() method (todo: wait with timeout in yield)
+				op.enqueueMail();
+				op.switchState(CLOSED);
+			}
+		},
+		CLOSED {
+			@Override
+			public boolean prepareToProcessRecord(ContinuousFileReaderOperator<?> op) {
+				LOG.warn("not processing any records while closed");
+				return false;
+			}
+		};
+
+		private static final Set<ReaderState> ACCEPT_SPLITS = EnumSet.of(IDLE, OPENING, READING);
+		/**
+		 * Possible transition FROM each state.
+		 */
+		private static final Map<ReaderState, Set<ReaderState>> TRANSITIONS;
+		static {
+			Map<ReaderState, Set<ReaderState>> tmpTransitions = new HashMap<>();
+			tmpTransitions.put(IDLE, EnumSet.of(OPENING, CLOSED));
+			tmpTransitions.put(OPENING, EnumSet.of(READING, CLOSING));
+			tmpTransitions.put(READING, EnumSet.of(IDLE, OPENING, CLOSING));
+			tmpTransitions.put(CLOSING, EnumSet.of(CLOSED));
+			tmpTransitions.put(CLOSED, EnumSet.noneOf(ReaderState.class));
+			TRANSITIONS = new EnumMap<>(tmpTransitions);
+		}
+
+		public boolean isAcceptingSplits() {
+			return ACCEPT_SPLITS.contains(this);
+		}
+
+		public final boolean isTerminal() {
+			return this == CLOSED;
+		}
+
+		public boolean canSwitchTo(ReaderState next) {
+			return TRANSITIONS
+					.getOrDefault(this, EnumSet.noneOf(ReaderState.class))
+					.contains(next);
+		}
```

Comment:
Could be replaced by passing that information to the constructors of the Enums.
`ReaderState(boolean acceptingSplits, boolean isTerminal, ReaderState... allowedTransitions)`
Transitions may introduce cyclic dependencies, so not 100% sure if this part is indeed working.

Reply:
Constructors wouldn't work here exactly because of cyclic dependencies.